### ISSUE TITLE
Miscompilation: comparing a pointer against all possible addresses returns `false`

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#130388]
Consider this code:
```rust
#![feature(strict_provenance)]

use std::ptr;

#[no_mangle]
pub fn test(x: *mut u8) -> bool {
    let local = 0u8;
    for i in 0..usize::MAX {
        if x.addr() == ptr::from_ref(&local).wrapping_add(i).addr() {
            return true;
        }
    }
    false
}
```

This compares `x.addr()` with *every possible address there is*. And yet, the optimized LLVM IR for this is:
```
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
define noundef zeroext i1 @test(ptr nocapture noundef readnone %x) unnamed_addr #0 {
start:
  ret i1 false
}
```

That's clearly nonsense.

This is the Rust version of https://github.com/llvm/llvm-project/issues/34450.

Cc @nikic @rust-lang/opsem

**Comment by orlp** [comment]
This also gets optimized to simply returning `false`:
```rust
#![feature(strict_provenance)]

use std::ptr;

#[no_mangle]
pub fn test(p: *mut u8) -> bool {
    let a = 0u8;
    p.addr() == ptr::from_ref(&a).addr()
}
```

**Comment by RalfJung** [comment]
@orlp yes but that can be justified. You are basically comparing an arbitrary non-deterministic value with `p` here, and we can rationalize this by saying that we can always put `a` at an address that's different from `p`, so we can optimize this to `false`. (Of course it is important that `a` then really does not actually live at address `p` -- if we can get contradictions that way, that would still be a bug. But I am not aware of a way to get contradictions that way.)

And this has nothing to do with strict provenance indeed, I just used `addr()` as it is shorter than `as usze` and has more well-defined semantics. ;)

**Comment by orlp** [comment]
> But I am not aware of a way to get contradictions here.

Simply put the `test` function into a loop that once again tests every possible address `x`. Unless the compiler really inserts some funky intentional-stack-misaligning logic into the preamble of `test`, it *must* be wrong at least once.

**Comment by RalfJung** [comment]
@orlp that doesn't work since `a`, logically, has a different address each invocation.

**Comment by orlp** [comment]
@RalfJung  It took some work, but here's a contradiction:
```rust
#[repr(C)]
struct Foo {
    a: u8,
    b: u8,
    c: u8,
}

#[no_mangle]
pub fn contradiction(x: *const u8) -> usize {
    let a = Foo { a: 0, b: 0, c: 0 };
    let ux = x as usize;
    let ua = &a.a as *const u8 as usize;
    let ub = &a.b as *const u8 as usize;
    let uc = &a.c as *const u8 as usize;

    if ua < ux && ux < uc {
        (ux != ub) as usize
    } else {
        1
    }
}
```

This compiles to just always returning 1, however, it's not possible for `ux` to lie between `ua` and `uc` yet still be inequal to `ub`.

**Comment by orlp** [comment]
Alright, I tried a bit to make a contradiction under those conditions and couldn't make one.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23130388.20Miscompilation.3A.20comparing.20a.20pointer.20against.20all.20pos.E2.80.A6)).

@rustbot label -I-prioritize +P-low

**Comment by nbdd0121** [comment]
I think this is another manifest of #107975 (adapted from `as` to `.addr()`: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=4365daa75e5b4d040e8cb88a317a8622) and https://github.com/llvm/llvm-project/issues/45725?

Essentially LLVM performs pointer comparision with provenance.

**Comment by nikic** [comment]
No, this is a completely unrelated issue. The issue description already links the corresponding LLVM issue.
